code/modules/atmospherics/gasmixtures/reactions.dm

```diff
@@ -212,113 +212,90 @@
-			SSresearch.science_tech.add_point_type(TECHWEB_POINT_TYPE_DEFAULT, 30000)//The science is cool though.
-			playsound(location, 'sound/effects/supermatter.ogg', 100, 0)
-		else
-			playsound(location, 'sound/effects/phasein.ogg', 75, 0)
-		//These will always happen, so be prepared
-		tesla_zap(location, zap_range, zap_power, TESLA_FUSION_FLAGS) //larpers beware
-		location.fire_nuclear_particles(power_ratio) //see code/modules/projectile/energy/nuclear_particle.dm
 
+	cached_gases[/datum/gas/plasma][MOLES] = plasma*scale_factor + FUSION_MOLE_THRESHOLD //Scales the gases back up
+	cached_gases[/datum/gas/carbon_dioxide][MOLES] = carbon*scale_factor + FUSION_MOLE_THRESHOLD
+	var/delta_plasma = initial_plasma - cached_gases[/datum/gas/plasma][MOLES]
+
+	reaction_energy += delta_plasma*PLASMA_BINDING_ENERGY //Energy is gained or lost corresponding to the creation or destruction of mass.
+	if(instability < FUSION_INSTABILITY_ENDOTHERMALITY)
+		reaction_energy = max(reaction_energy,0) //Stable reactions don't end up endothermic.
+	else if (reaction_energy < 0)
+		reaction_energy *= (instability-FUSION_INSTABILITY_ENDOTHERMALITY)**0.5
+
+	if(air.thermal_energy() + reaction_energy < 0) //No using energy that doesn't exist.
+		cached_gases[/datum/gas/plasma][MOLES] = initial_plasma
+		cached_gases[/datum/gas/carbon_dioxide][MOLES] = initial_carbon
+		return NO_REACTION
+	cached_gases[/datum/gas/tritium][MOLES] -= FUSION_TRITIUM_MOLES_USED
+	//The decay of the tritium and the reaction's energy produces waste gases, different ones depending on whether the reaction is endo or exothermic
 	if(reaction_energy > 0)
+		air.assert_gases(/datum/gas/oxygen,/datum/gas/nitrous_oxide)
+		cached_gases[/datum/gas/oxygen][MOLES] += FUSION_TRITIUM_MOLES_USED*(reaction_energy*FUSION_TRITIUM_CONVERSION_COEFFICIENT)
+		cached_gases[/datum/gas/nitrous_oxide][MOLES] += FUSION_TRITIUM_MOLES_USED*(reaction_energy*FUSION_TRITIUM_CONVERSION_COEFFICIENT)
+	else
+		air.assert_gases(/datum/gas/bz,/datum/gas/nitryl)
+		cached_gases[/datum/gas/bz][MOLES] += FUSION_TRITIUM_MOLES_USED*(reaction_energy*-FUSION_TRITIUM_CONVERSION_COEFFICIENT)
+		cached_gases[/datum/gas/nitryl][MOLES] += FUSION_TRITIUM_MOLES_USED*(reaction_energy*-FUSION_TRITIUM_CONVERSION_COEFFICIENT)
+
+	if(reaction_energy)
+		if(location)
+			var/particle_chance = ((PARTICLE_CHANCE_CONSTANT)/(reaction_energy-PARTICLE_CHANCE_CONSTANT)) + 1//Asymptopically approaches 100% as the energy of the reaction goes up.
+			if(prob(PERCENT(particle_chance)))
+				location.fire_nuclear_particle()
+			var/rad_power = max((FUSION_RAD_COEFFICIENT/instability) + FUSION_RAD_MAX,0)
+			radiation_pulse(location,rad_power)
+
 		var/new_heat_capacity = air.heat_capacity()
 		if(new_heat_capacity > MINIMUM_HEAT_CAPACITY)
-			air.temperature = max(((temperature*old_heat_capacity + reaction_energy)/new_heat_capacity),TCMB)
+			air.temperature = CLAMP(((air.temperature*old_heat_capacity + reaction_energy)/new_heat_capacity),TCMB,INFINITY)
```

Comment:
This feels like there's something wrong with this setup if this is necessary ?

Reply:
I'm not a huge fan of doing it either, it's just kind of an unfortunate consequence of the fact that the way it scales, for a really large reactor(Lots of mols) doing powerful fusion(High temperatures), it can actually hit the float overflow point after a very long time(Like 4+ hours). I don't think it's something that will ever come up naturally in game, but it was able to happen in earlier stages of testing and I'd rather play it safe than sorry.